### ISSUE TITLE
ClassCastException: PropertyValueLiteral cannot be cast to class PropertyKey

### ISSUE THREAD

**Issue by fbricon** [eclipse/lsp4mp#100]
Steps to reproduce:

clone https://github.com/fbricon/quarkus-spread-poll
follow  https://www.youtube.com/watch?v=0KYL6Q1QO4c&feature=youtu.be&t=1125

```
Nov 02, 2020 3:18:59 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
java.util.concurrent.CompletionException: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1238)
	at java.base/java.util.concurrent.CompletableFuture$BiApply.tryFire(CompletableFuture.java:1205)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActionsForUnknown(MicroProfileCodeActions.java:117)
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActions(MicroProfileCodeActions.java:88)
	at org.eclipse.lsp4mp.services.MicroProfileLanguageService.doCodeActions(MicroProfileLanguageService.java:229)
	at org.eclipse.lsp4mp.ls.ApplicationPropertiesTextDocumentService.lambda$codeAction$13(ApplicationPropertiesTextDocumentService.java:259)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1236)
	... 7 more

[Error - 3:18:59 PM] Request textDocument/codeAction failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1238)
	at java.base/java.util.concurrent.CompletableFuture$BiApply.tryFire(CompletableFuture.java:1205)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActionsForUnknown(MicroProfileCodeActions.java:117)
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActions(MicroProfileCodeActions.java:88)
	at org.eclipse.lsp4mp.services.MicroProfileLanguageService.doCodeActions(MicroProfileLanguageService.java:229)
	at org.eclipse.lsp4mp.ls.ApplicationPropertiesTextDocumentService.lambda$codeAction$13(ApplicationPropertiesTextDocumentService.java:259)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1236)
	... 7 more

Nov 02, 2020 3:19:02 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
java.util.concurrent.CompletionException: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1238)
	at java.base/java.util.concurrent.CompletableFuture$BiApply.tryFire(CompletableFuture.java:1205)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActionsForUnknown(MicroProfileCodeActions.java:117)
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActions(MicroProfileCodeActions.java:88)
	at org.eclipse.lsp4mp.services.MicroProfileLanguageService.doCodeActions(MicroProfileLanguageService.java:229)
	at org.eclipse.lsp4mp.ls.ApplicationPropertiesTextDocumentService.lambda$codeAction$13(ApplicationPropertiesTextDocumentService.java:259)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1236)
	... 7 more

[Error - 3:19:02 PM] Request textDocument/codeAction failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1238)
	at java.base/java.util.concurrent.CompletableFuture$BiApply.tryFire(CompletableFuture.java:1205)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActionsForUnknown(MicroProfileCodeActions.java:117)
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActions(MicroProfileCodeActions.java:88)
	at org.eclipse.lsp4mp.services.MicroProfileLanguageService.doCodeActions(MicroProfileLanguageService.java:229)
	at org.eclipse.lsp4mp.ls.ApplicationPropertiesTextDocumentService.lambda$codeAction$13(ApplicationPropertiesTextDocumentService.java:259)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1236)
	... 7 more

Nov 02, 2020 3:21:28 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
java.util.concurrent.CompletionException: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1238)
	at java.base/java.util.concurrent.CompletableFuture$BiApply.tryFire(CompletableFuture.java:1205)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActionsForUnknown(MicroProfileCodeActions.java:117)
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActions(MicroProfileCodeActions.java:88)
	at org.eclipse.lsp4mp.services.MicroProfileLanguageService.doCodeActions(MicroProfileLanguageService.java:229)
	at org.eclipse.lsp4mp.ls.ApplicationPropertiesTextDocumentService.lambda$codeAction$13(ApplicationPropertiesTextDocumentService.java:259)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1236)
	... 7 more

[Error - 3:21:28 PM] Request textDocument/codeAction failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1238)
	at java.base/java.util.concurrent.CompletableFuture$BiApply.tryFire(CompletableFuture.java:1205)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActionsForUnknown(MicroProfileCodeActions.java:117)
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActions(MicroProfileCodeActions.java:88)
	at org.eclipse.lsp4mp.services.MicroProfileLanguageService.doCodeActions(MicroProfileLanguageService.java:229)
	at org.eclipse.lsp4mp.ls.ApplicationPropertiesTextDocumentService.lambda$codeAction$13(ApplicationPropertiesTextDocumentService.java:259)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1236)
	... 7 more

Nov 02, 2020 3:21:40 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
java.util.concurrent.CompletionException: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1238)
	at java.base/java.util.concurrent.CompletableFuture$BiApply.tryFire(CompletableFuture.java:1205)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActionsForUnknown(MicroProfileCodeActions.java:117)
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActions(MicroProfileCodeActions.java:88)
	at org.eclipse.lsp4mp.services.MicroProfileLanguageService.doCodeActions(MicroProfileLanguageService.java:229)
	at org.eclipse.lsp4mp.ls.ApplicationPropertiesTextDocumentService.lambda$codeAction$13(ApplicationPropertiesTextDocumentService.java:259)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1236)
	... 7 more

[Error - 3:21:40 PM] Request textDocument/codeAction failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1238)
	at java.base/java.util.concurrent.CompletableFuture$BiApply.tryFire(CompletableFuture.java:1205)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.PropertyValueLiteral cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.PropertyValueLiteral and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActionsForUnknown(MicroProfileCodeActions.java:117)
	at org.eclipse.lsp4mp.services.MicroProfileCodeActions.doCodeActions(MicroProfileCodeActions.java:88)
	at org.eclipse.lsp4mp.services.MicroProfileLanguageService.doCodeActions(MicroProfileLanguageService.java:229)
	at org.eclipse.lsp4mp.ls.ApplicationPropertiesTextDocumentService.lambda$codeAction$13(ApplicationPropertiesTextDocumentService.java:259)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1236)
	... 7 more
```

**Comment by fbricon** [comment]
I can't reproduce the issue. Didn't see either before performing the demo live.

**Comment by angelozerr** [comment]
The error is very strange, because we create a diagnostic for property key for unknown validation type and code action retrieve the property key from this diagnostic range. I don't understand how we could have this usecase?

**Comment by datho7561** [comment]
I can take a look at this

**Comment by datho7561** [comment]
Having trouble reproducing, but I found an infinite loop in the parser

**Comment by datho7561** [comment]
I managed to reproduce it. Here are the steps, do them as soon as the language server starts:
1. use vscode-quarkus
2. use Fred's project
3. delete code connecting postgres in application.properties
4. activate `qds` snippet
5. select `postgres` for database type then tab to next property
6. as the databse username, enter `quarkus=-=`

Logs:
```
Params: {
    "textDocument": {
        "uri": "file:///home/dthompson/Documents/TestProjects/quarkus-spread-poll/src/main/resources/application.properties"
    },
    "range": {
        "start": {
            "line": 13,
            "character": 38
        },
        "end": {
            "line": 14,
            "character": 26
        }
    },
    "context": {
        "diagnostics": [
            {
                "range": {
                    "start": {
                        "line": 13,
                        "character": 38
                    },
                    "end": {
                        "line": 14,
                        "character": 26
                    }
                },
                "message": "Unknown property '\nquarkus.datasource.passwor'",
                "severity": 2,
                "code": "unknown",
                "source": "microprofile"
            }
        ]
    }
}
```

I think it might be adding an `=-=` that causes the parser to not have a good time; I will continue to investigate and see how often I can reproduce this way

**Comment by datho7561** [comment]
Trying under cpulimit...

**Comment by datho7561** [comment]
Okay. This is reproducible reliably:
0. Install vscode-microprofile (`vsce package`, then install from the vsix)
1. Clone Fred's project
2. Run vscode-quarkus in extension debug mode
3. Open the project as a folder workspace
4. If you have opened the project before, clean the Java workspace, and fully relaunch the extension debugging
5. Use `cpulimit` or a similar tool to limit microprofile-ls's access to the CPU (a limit of 3% worked for me)
6. Open `application.properties`
7. Delete the existing database connection configuration
8. use the `qds` snippet to generate a new database connection configuration
9. Select `postgres` as the type, then tab to the next config property
10. Enter `quarkus=-=`
11. Tab to the next entry
12. Crash (may need to wait a few seconds for the crash due to the cpu restrictions)

Don't know the root cause yet.

**Comment by angelozerr** [comment]
The thing I don't understand is that fred application.properties have no `${` and your usecase have no `${`

My question is why we have a PropertyValueExpression in this case?

**Comment by datho7561** [comment]
PropertyValueLiteral is used regardless of if there are `${}` or not. We have this structure in the model:

```
           Property
           /            \
PropertyKey    PropertyValue
                                |
                       1 or more PropertyValueLiteral or PropertyValueExpression
```

this is so that we can read something like this and represent it in the model:

```
endpoint = https://localhost:${port}/my/api
```

**Comment by angelozerr** [comment]
@datho7561 thanks for your clarification, I confused PropertyValueLiteral with PropertyValueExpression. Thanks for the clarification.

**Comment by datho7561** [comment]
Can you think of a better name for it? If the name is easy to confuse, we should come up with a better one.

Going to write unit and integration tests to see if I can pinpoint the problem that way.

**Comment by angelozerr** [comment]
No no the name is perfect! It's my fault, I wrote my comments too quickly, sorry -(

**Comment by datho7561** [comment]
I suspect that this issue is related to concurrency, i.e. the code that generates the `CodeAction` has a DOM model of the properties file that is out of sync with the actual properties file. It has been hard to debug.

**Comment by datho7561** [comment]
Maybe this was related to #272. I will try to reproduce this bug to see if it got fixed.

**Comment by rgrunber** [comment]
> Maybe this was related to #272. I will try to reproduce this bug to see if it got fixed.

Yes! I don't think I ever posted the stacktrace, because once we found the problem, it seemed pointless, but here is the original stacktrace I sent from that issue :

```
Jul. 08, 2022 11:49:44 A.M. org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.Assign cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.Assign and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
java.util.concurrent.CompletionException: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.Assign cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.Assign and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.ClassCastException: class org.eclipse.lsp4mp.model.Assign cannot be cast to class org.eclipse.lsp4mp.model.PropertyKey (org.eclipse.lsp4mp.model.Assign and org.eclipse.lsp4mp.model.PropertyKey are in unnamed module of loader 'app')
	at org.eclipse.lsp4mp.services.properties.PropertiesFileCodeActions.doCodeActionsForUnknown(PropertiesFileCodeActions.java:115)
	at org.eclipse.lsp4mp.services.properties.PropertiesFileCodeActions.doCodeActions(PropertiesFileCodeActions.java:87)
	at org.eclipse.lsp4mp.services.properties.PropertiesFileLanguageService.doCodeActions(PropertiesFileLanguageService.java:227)
	at org.eclipse.lsp4mp.ls.properties.PropertiesFileTextDocumentService.lambda$codeAction$12(PropertiesFileTextDocumentService.java:228)
	at org.eclipse.lsp4mp.ls.commons.ModelTextDocuments.lambda$computeModelAsync$0(ModelTextDocuments.java:119)
	... 7 more
```

Feel free to close this.

**Comment by datho7561** [comment]
Closed by #273
